File: Haskell/2023/09.hs
import Lib

type Input = [[Int]]

main :: IO ()
main = aoc 2023 9 setup solve1 solve2 ["1"]

solve1 :: Input -> Int
solve1 = sum . map solve

solve2 :: Input -> Int
solve2 = sum . map (solve . reverse)

solve :: [Int] -> Int
solve = sum . map last . takeWhile (any (/= 0)) . iterate step

step :: [Int] -> [Int]
step = zipWith (-) =<< tail

setup :: String -> Input
setup = map (map read . words) . lines

File: Rust/2023/09.rs
#![feature(test)]

type Input = Vec<Vec<i32>>;

fn setup(input: &str) -> Input {
    input
        .lines()
        .map(|line| {
            line.split_whitespace()
                .map(|x| x.parse().unwrap())
                .collect()
        })
        .collect()
}

fn solve(mut nums: Vec<i32>) -> i32 {
    for j in 1.. {
        let mut all_zero = true;
        let mut prev = nums[j - 1];
        for i in nums.iter_mut().skip(j) {
            *i = prev - *i;
            prev -= *i;

            all_zero &= *i == 0;
        }

        if all_zero {
            break;
        }
    }

    nums.into_iter().sum()
}

fn part1(input: &Input) -> i32 {
    input
        .clone()
        .into_iter()
        .map(|mut nums| {
            nums.reverse();
            solve(nums)
        })
        .sum()
}

fn part2(input: &Input) -> i32 {
    input.clone().into_iter().map(solve).sum()
}

aoc::main!(2023, 9, ex: 1);
